src/jit/gschecks.cpp

```diff
@@ -423,8 +423,11 @@ void Compiler::gsParamsToShadows()
         lvaTable[shadowVar].lvLclFieldExpr     = varDsc->lvLclFieldExpr;
         lvaTable[shadowVar].lvLiveAcrossUCall  = varDsc->lvLiveAcrossUCall;
 #endif
-        lvaTable[shadowVar].lvVerTypeInfo    = varDsc->lvVerTypeInfo;
-        lvaTable[shadowVar].lvGcLayout       = varDsc->lvGcLayout;
+        lvaTable[shadowVar].lvVerTypeInfo = varDsc->lvVerTypeInfo;
+        if (varTypeIsStruct(type))
+        {
+            lvaTable[shadowVar].SetLayout(varDsc->GetLayout());
```

Comment:
It looks like this should not be needed, `gsParamsToShadows` calls `lvaSetStruct` so both `lvVerTypeInfo` and the layout would be set anyway. Question is why isn't `lvaSetStruct` simply called here, makes no sense to me to delay it.

Reply:
I"m a bit confused by your comment. You said that it calls `lvaSetStruct`, but it doesn't seem to. Did you mean that it *should* do so?

Reply:
It does call it but in a different place:
* first it creates the variables, without calling `lvaSetStruct`
* then it changes the IR to use the new variables (ugly because the IR will now temporarily be in an invalid state - the newly created variables may have not been fully setup without `lvaSetStruct`)
* then it generates copies from the old variables to the new variables and here it also calls `lvaSetStruct` - https://github.com/dotnet/coreclr/blob/9479f67577bbb02ea611777b00308f42252fb2bc/src/jit/gschecks.cpp#L523-L527
I don't see any reason not to call `lvaSetStruct` when the variables are created.

But this code is ancient. There are still traces of old struct handling - it wraps the lclvar nodes in `ADDR` nodes and then calls `gtNewCpObjNode`. Which will of course remove the `ADDR` nodes as they're not necessary.

A bunch of extra complication basically.

Reply:
Thanks - I didn't look down far enough. It should probably be cleaned up. It's not clear how much it matters, but I agree that it's unwise to have the inconsistent state, even temporarily.

Reply:
Opened https://github.com/dotnet/coreclr/issues/26550